### DISCUSSION TITLE
Modules and scattered definitions

### DISCUSSION THREAD

**Discussion by algr** [rems-project/sail#404] [Q&A] [answered]
The normal rules on scoping seem to imply that if B requires A and C requires B, but not A, then C will see B's definitions but not A's. Experiement suggests that C will see the names of scattered definitions introduced in A - and it can add to those definitions. So if A has something like
```
scattered enum E
enum clause E = E_one
```
B has
```
enum clause E = E_two
```
C has
```
enum clause E = E_three
```
Sail will accept this, and references to `E` in C will refer to the type name introduced in A. References to `E_one` in C will fail but Sail will helpfully point out that it is defined in A and will not allow it to be reused in C.

So it seems that scatter-defined names (is that the right terminology?) propagate through all imports, even indirect ones, in a way that normally defined names don't. I looked through the documentation but couldn't see this mentioned anywhere.

**Comment by Alasdair** [comment, marked as answer]
Should be fixed by https://github.com/rems-project/sail/pull/406

Currently the module system is WIP, so there are probably some similar issues. The in-scope check is done when accessing items from a typing context, so I probably need to go through and check all functions that are modifying things in the typing context to ensure they are not doing this in a way that breaks the rules.

**Reply by algr** [reply]
Thanks - I see "Enumeration E is not in scope" now. So that is consistent. I wasn't sure whether the original behavior was intentional.
